File: src/data/workspace.ts
import { existsSync, readFileSync, readdirSync, statSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';
import type { OcWorkspaceFile } from '../types.js';

const WORKSPACE = join(homedir(), '.openclaw', 'workspace');

const WORKSPACE_FILES = [
  { file: 'IDENTITY.md', desc: 'Name, pronouns, emoji, avatar' },
  { file: 'SOUL.md',     desc: 'Core personality and values' },
  { file: 'USER.md',     desc: "Dave's profile and preferences" },
  { file: 'AGENTS.md',   desc: 'Operating rules and workspace guide' },
  { file: 'TOOLS.md',    desc: 'Environment-specific cheat sheet' },
  { file: 'MEMORY.md',   desc: 'Curated long-term memory' },
  { file: 'HEARTBEAT.md',desc: 'Active heartbeat checklist' },
  { file: 'BOOTSTRAP.md',desc: 'First-run bootstrap instructions' },
];

function readWorkspaceFile(filePath: string, id: string, name: string): OcWorkspaceFile | null {
  try {
    const content = readFileSync(filePath, 'utf-8');
    const stat = statSync(filePath);
    return {
      kind: 'workspace',
      id,
      name,
      filePath,
      preview: content.slice(0, 500),
      fullContent: content,
      wordCount: content.split(/\s+/).filter(Boolean).length,
      lastModified: stat.mtime.toISOString(),
    };
  } catch {
    return null;
  }
}

export function loadWorkspace(): OcWorkspaceFile[] {
  const results: OcWorkspaceFile[] = [];

  for (const { file } of WORKSPACE_FILES) {
    const filePath = join(WORKSPACE, file);
    if (!existsSync(filePath)) continue;
    const item = readWorkspaceFile(filePath, `workspace#${file}`, file);
    if (item) results.push(item);
  }

  const memDir = join(WORKSPACE, 'memory');
  if (existsSync(memDir)) {
    const memFiles = readdirSync(memDir)
      .filter(f => f.endsWith('.md'))
      .sort()
      .reverse();
    for (const file of memFiles) {
      const filePath = join(memDir, file);
      const item = readWorkspaceFile(filePath, `workspace#memory/${file}`, `memory/${file}`);
      if (item) results.push(item);
    }
  }

  return results;
}

File: internal/data/workspace.go
package data

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
)

var workspaceFiles = []string{
	"IDENTITY.md",
	"SOUL.md",
	"USER.md",
	"AGENTS.md",
	"TOOLS.md",
	"MEMORY.md",
	"HEARTBEAT.md",
	"BOOTSTRAP.md",
}

func readWorkspaceFile(filePath, id, name string) *OcWorkspaceFile {
	b, err := os.ReadFile(filePath)
	if err != nil {
		return nil
	}
	info, err := os.Stat(filePath)
	if err != nil {
		return nil
	}
	content := string(b)
	preview := content
	if len(preview) > 500 {
		preview = preview[:500]
	}
	wordCount := len(strings.Fields(content))
	return &OcWorkspaceFile{
		ID:           id,
		Name_:        name,
		FilePath:     filePath,
		Preview:      preview,
		FullContent:  content,
		WordCount:    wordCount,
		LastModified: info.ModTime().Format("2006-01-02T15:04:05Z07:00"),
	}
}

func LoadWorkspace() []OcWorkspaceFile {
	home, _ := os.UserHomeDir()
	workspace := filepath.Join(home, ".openclaw", "workspace")

	var items []OcWorkspaceFile
	for _, file := range workspaceFiles {
		path := filepath.Join(workspace, file)
		if _, err := os.Stat(path); err != nil {
			continue
		}
		if item := readWorkspaceFile(path, "workspace#"+file, file); item != nil {
			items = append(items, *item)
		}
	}

	memDir := filepath.Join(workspace, "memory")
	if entries, err := os.ReadDir(memDir); err == nil {
		var memFiles []string
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".md") {
				memFiles = append(memFiles, e.Name())
			}
		}
		sort.Sort(sort.Reverse(sort.StringSlice(memFiles)))
		for _, file := range memFiles {
			path := filepath.Join(memDir, file)
			name := "memory/" + file
			if item := readWorkspaceFile(path, "workspace#"+name, name); item != nil {
				items = append(items, *item)
			}
		}
	}
	return items
}
